fmn/formatters.py

```diff
@@ -403,78 +415,76 @@ def email_batch(messages, recipient):
     if len(messages) == 1:
         return email(messages[0], recipient)
 
-    def _format_line(msg):
-        timestamp = datetime.datetime.fromtimestamp(msg['timestamp'])
-        link = fedmsg.meta.msg2link(msg, **config.app_conf) or u''
-        payload = fedmsg.meta.msg2subtitle(msg, **config.app_conf) or u''
+    email_message = _base_email()
+    email_message.add_header('To', recipient['email address'].encode('utf-8'))
+    email_message.add_header('From', config.app_conf['fmn.email.from_address'].encode('utf-8'))
+    email_message.add_header(
+        'Subject', u'Fedora Notifications Digest ({n} updates)'.format(n=len(messages)))
+    try:
+        topics = set([message['topic'] for message in messages])
+        categories = set([topic.split('.')[3] for topic in topics])
+        for topic in topics:
+            email_message.add_header('X-Fedmsg-Topic', topic.encode('utf-8'))
+        for cat in categories:
+            email_message.add_header('X-Fedmsg-Category', cat.encode('utf-8'))
+    except Exception:
+        _log.exception('Unable to parse message topic and category for %r', messages)
 
-        if recipient.get('verbose', True):
-            longform = fedmsg.meta.msg2long_form(msg, **config.app_conf) or u''
-            if longform:
-                payload += "\n" + longform
-
-        return timestamp.strftime("%c") + ", " + payload + "\n\t" + link
-
-    from_address = config.app_conf['fmn.email.from_address']
-    topics = set([message['topic'] for message in messages])
-    categories = set([topic.split('.')[3] for topic in topics])
-
-    def squash(items):
-        return reduce(set.union, items, set())
-    usernames = squash([
-        fedmsg.meta.msg2usernames(msg, **config.app_conf)
-        for msg in messages])
-    packages = squash([
-        fedmsg.meta.msg2packages(msg, **config.app_conf)
-        for msg in messages])
-    subject = u'Fedora Notifications Digest ({n} updates)'.format(n=len(messages))
-    summary = u"Digest summary:\n"
-    i = 0
     for msg in messages:
-        i = i + 1
-        line = fedmsg.meta.msg2subtitle(msg, **config.app_conf) or u''
-        summary += '%d.\t%s\n' % (i, line)
+        try:
+            for username in fedmsg.meta.msg2usernames(msg, **config.app_conf) or []:
```

Comment:
Line 369 we have the same block maybe this could be put inside a function to avoid duplication

Reply:
Ah, I see what you mean. I think I can factor out a lot of code with some tweaks to the _base_email() function.